src/Nethermind/Nethermind.Optimism/OPL1CostHelper.cs

```diff
@@ -37,11 +44,32 @@ public UInt256 ComputeL1Cost(Transaction tx, BlockHeader header, IWorldState wor
         if (dataGas == 0)
             return UInt256.Zero;
 
-        UInt256 l1BaseFee = new(worldState.Get(_l1BaseFeeSlot), true);
-        UInt256 overhead = new(worldState.Get(_overheadSlot), true);
-        UInt256 scalar = new(worldState.Get(_scalarSlot), true);
 
-        return ((UInt256)dataGas + overhead) * l1BaseFee * scalar / 1_000_000;
+        if (_opConfigHelper.IsEcotone(header))
+        {
+            // Ecotone formula: (dataGas) * (16*l1BaseFee*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar) / 16e6
+            UInt256 l1BaseFee = new(worldState.Get(_l1BaseFeeSlot), true);
+            UInt256 blobBaseFee = new(worldState.Get(_blobBaseFeeSlot), true);
+
+            ReadOnlySpan<byte> scalarData = worldState.Get(_baseFeeScalarSlot);
+            Span<byte> scalarDataAligned = stackalloc byte[32];
+            scalarData.CopyTo(scalarDataAligned[(32 - scalarData.Length)..]);
+
+            const int offset = 32 - 12 - 4;
+            UInt256 l1BaseFeeScalar = new(scalarDataAligned[offset..(offset + 4)], true);
+            UInt256 l1BlobBaseFeeScalar = new(scalarDataAligned[(offset + 4)..(offset + 8)], true);
+
+            return (UInt256)dataGas * (16 * l1BaseFee * l1BaseFeeScalar + blobBaseFee * l1BlobBaseFeeScalar) /
+                   16_000_000;
+        }
+        else
+        {
+            UInt256 l1BaseFee = new(worldState.Get(_l1BaseFeeSlot), true);
+            UInt256 overhead = new(worldState.Get(_overheadSlot), true);
+            UInt256 scalar = new(worldState.Get(_scalarSlot), true);
```

Comment:
potentially `l1BaseFee` and `scalar` could go before `if` statement as they are used in both paths.